### ISSUE TITLE
Ensure the client can't trigger Sentry rate limiting

### ISSUE THREAD

**Issue by seanh** [hypothesis/client#759]
**Fix the client to never trigger Sentry rate limiting**.

As things stand at the time of writing [client Sentry reporting has been disabled](https://hypothes-is.slack.com/archives/C4K6M7P5E/p1534262792000100) so it's no longer triggering Sentry's organization-wide rate limiting and affecting other projects. This issue is to fix it so that the client can't flood Sentry and trigger rate limiting, so that we can turn on client Sentry reporting again. _Or_, we could decide not to use Sentry for the client anymore (maybe setting up another tool instead) and then close this issue.

Please note that this issue is _not about improving client Sentry signal-noise ratio_. This issue is just about **preventing the client from triggering Sentry rate limiting** by any means. For example we might decide not to use Sentry for the client anymore, if it isn't an appropriate tool. The low signal-noise ratio of the client's Sentry events is a problem that causes actual client issues to be missed due to being flooded by noise issues, and depending on the solution we _might_ end up fixing the signal-noise as part of making sure it can't trigger rate limiting, but the signal-noise is a separate issue.

Also please note that this issue is _not about fixing a particular client issue or issues that have caused a lot of Sentry events recently_. As I hope I've documented thoroughly below, a low signal-noise ratio and flooding of Sentry by frontend code (mainly the client) has been an ongoing issue for years. A number of times in the past specific issues that have caused floods of events have been fixed, only for it to happen again later because of another issue. It looks like some attempts at general solutions by filtering or whitelisting have been made but also not succeeded at fixing the problem for good. We may need some new thinking here.

## Done when

- [ ] Changes have been made (e.g. in the client, or in Sentry) such that we don't think the client can flood Sentry enough to trigger organization-wide rate limits any more, and Sentry reporting has been re-enabled for the client. **Or** when we have decided not to use Sentry for the client anymore, removed Sentry support from the client, and perhaps set up an alternative tool.

## Notes

In production the client reports far too many events to Sentry for it to be useful (many tens of thousands of events per week, signal to noise ratio is too low). Given the volume of client data in Sentry - 712 unresolved issues, hundreds of thousands of events - it doesn't appear that we're making effective use of this data. This has even sometimes caused us to hit our organization-wide monthly cap / rate limit, causing events from other projects (that we _are_ trying to make use of) to be blocked by Sentry's rate limiting. In the case of the lms app, it reported a mere 33 events in one week of which 15 (45%) were blocked by rate limiting triggered by the client, This really hampers our ability to know what's going on with lms users who're seeing problems.

## See also

* A bunch of recent Slack threads:
  * https://hypothes-is.slack.com/archives/C4K6M7P5E/p1534181246000287
  * https://hypothes-is.slack.com/archives/C4K6M7P5E/p1534241622000100
  * https://hypothes-is.slack.com/archives/C4K6M7P5E/p1534241941000200
  * https://hypothes-is.slack.com/archives/C4K6M7P5E/p1534242976000100
  * https://hypothes-is.slack.com/archives/C4K6M7P5E/p1534243430000100
  * https://hypothes-is.slack.com/archives/C4K6M7P5E/p1534262792000100

* A [previous recent issue that caused the client to hit Sentry's cap](https://github.com/hypothesis/client/issues/750)

* [h feature that avoids sending certain events to Sentry](https://github.com/hypothesis/h/pull/4742)

* Open issue from Apr 2017 [Better alerting for client Sentry reports #276](https://github.com/hypothesis/product-backlog/issues/276#issuecomment-296740556)

* Open issue from Apr 2017 [Implement client-side rate limiting of error reports #277](https://github.com/hypothesis/product-backlog/issues/277)

* A previous (Sep 2016) [fix for a client bug that caused it to deluge Sentry](https://github.com/hypothesis/client/pull/138)

* The client flooding Sentry with high volumes of events that both drown out any signal (any actual client bugs we should care about) and block Sentry reporting by other projects has been an ongoing problem for a long time.

  From a quick Slack search it looks like there were major Sentry flooding issues in at least Aug 2018, Apr 2017, Nov, Sep & Aug 2016 and Mar 2016. Generally speaking I don't think we've made effective use of the client's Sentry events in the ~2.5 years this Sentry reporting was enabled for the client - even when it wasn't completely flooding Sentry or triggering rate limiting that affected other projects, it has always been far too many events, far too low a signal-noise ration.

  * Apr 2017: [Client triggered Sentry's cap](https://hypothes-is.slack.com/archives/C1M8NH76X/p1493028735943862)
  * Nov 2016: ["we either need to fix it, or turn off Sentry reporting from the frontend code until we can"](https://hypothes-is.slack.com/archives/C0LUWQQJJ/p1479819422000086)
  * Sep 2016: ["The prod-client Sentry project is not really helping us at the moment. The volume of issues is far too high -- some of them important for us to look into and some of them far less so..."](https://trello.com/c/B6KN6YDC)
  * Sep 2016: [Common causes of client Sentry errors](https://trello.com/c/zS1a1IsH/674-sentry-common-client-errors)
  * Sep 2016: [Sentry error whitelisting](https://trello.com/c/ViNWkvQ8/113-sentry-setup-whitelists-for-web-service-javascript-code)
  * Sep 2016: [I'm really concerned about the signal to noise ratio in Sentry, as I've said before. there's a lot of junk in here, and as a result we're missing actual problems ](https://hypothes-is.slack.com/archives/C076LQFA4/p1473671433001293)
  * Aug 2016: [my point is a simple one: there is a bunch of noise at the moment, which means that we collectively are ignoring Sentry for the client. I can't remember the last time I saw someone do some gardening of front-end Sentry reports. so either we do something about it and get the signal-to-noise ratio to point where it's useful, or I'm turning it off. if we can't get the signal-to-noise ratio to a point where it's useful, I'm turning it off](https://hypothes-is.slack.com/archives/C076LQFA4/p1472549687000838)
  * Aug 2016: [I don't think we're ever going to get good use out of it unless we can filter out the vast majority of the noise. if receiving messages like that is unavoidable, then I'm not really sure Sentry is the right tool for the job](https://hypothes-is.slack.com/archives/C076LQFA4/p1472548718000815)
  * Aug 2016: [unfortunately the torrent of crap from the client appears to be negatively affecting the server project. I think we might have to disable sentry for the client entirely if we can't get a handle on the volume of crap it's sending](https://hypothes-is.slack.com/archives/C076LQFA4/p1471959635005155)
  * Mar 2016: [if we split client/web service sentry accounts, we can use this to ensure one doesn't swamp the other](https://hypothes-is.slack.com/archives/C076LQFA4/p1457459549005547) (note this comment refers to moving the client to a separate Sentry _account_ / organization, not just to a separate project, Sentry's limits are organization-wide not per-project)

**Comment by robertknight** [comment]
The issue in Via that was causing a ton of errors to be sent to Sentry has been resolved in https://github.com/hypothesis/via/pull/171. We might be able to re-enable Sentry for the client. It probably would still be a good idea to add some client-side rate limiting of events first though, so no one client instance can spam Sentry.

**Comment by robertknight** [comment]
Ongoing work here as part of switching the client to the new Sentry SDK: https://github.com/hypothesis/client/pull/1320

**Comment by robertknight** [comment]
This has now been addressed in various ways:

- The client now limits how many events it will send to Sentry in a given session. See https://github.com/hypothesis/client/pull/1320.
- We fixed the issues that were causing the majority of existing reports: https://github.com/hypothesis/client/issues/187

There is a remaining issue where transient network errors or faults in h (eg. there was a brief outage yesterday that caused a lot of 5xx responses) can result in many client error reports. This is being addressed in https://github.com/hypothesis/client/issues/3631.

On the basis of these changes, mainly the client-side rate limiting added in https://github.com/hypothesis/client/pull/1320, I'm going to close this for now.
